code/modules/projectiles/guns/energy.dm

```diff
@@ -594,3 +594,776 @@
+	var/radial_icon_state = "laser"
+
+/datum/energy_sniper_mode/standard
+	charge_cost = 50
+	ammo = /datum/ammo/energy/lasgun/marine/sniper
+	fire_delay = 1 SECONDS
+	fire_sound = 'sound/weapons/guns/fire/Laser Sniper Standard.ogg'
+	message_to_user = "You set the Laser rifle's charge mode to standard fire."
+	fire_mode = GUN_FIREMODE_AUTOMATIC
+	icon_state = "tes"
+
+/datum/energy_sniper_mode/heat
+	charge_cost = 150
+	ammo = /datum/ammo/energy/lasgun/marine/sniper_heat
+	fire_delay = 1.5 SECONDS
+	fire_sound = 'sound/weapons/guns/fire/laser3.ogg'
+	message_to_user = "You set the Laser Sniper's charge mode to wave heat."
+	fire_mode = GUN_FIREMODE_AUTOMATIC
+	icon_state = "tes"
+	radial_icon_state = "laser_heat"
+
+/obj/item/weapon/gun/energy/lasgun/standard_marine_sniper/unique_action(mob/user)
+	return switch_modes(user)
+
+/obj/item/weapon/gun/energy/lasgun/standard_marine_sniper/proc/switch_modes(mob/user)
+	if(!user)
+		CRASH("switch_modes called with no user.")
+
+	var/list/available_modes = list()
+	for(var/mode in mode_list)
+		available_modes += list("[mode]" = image(icon = initial(mode_list[mode].radial_icon), icon_state = initial(mode_list[mode].radial_icon_state)))
+
+	var/datum/energy_sniper_mode/choice = mode_list[show_radial_menu(user, user, available_modes, null, 64, tooltips = TRUE)]
+	if(!choice)
+		return
+
+
+	playsound(user, 'sound/weapons/emitter.ogg', 5, FALSE, 2)
+
+
+	gun_firemode = initial(choice.fire_mode)
+
+
+	charge_cost = initial(choice.charge_cost)
+	ammo = GLOB.ammo_list[initial(choice.ammo)]
+	fire_delay = initial(choice.fire_delay)
+	fire_sound = initial(choice.fire_sound)
+	SEND_SIGNAL(src, COMSIG_GUN_AUTOFIREDELAY_MODIFIED, fire_delay)
+	SEND_SIGNAL(src, COMSIG_GUN_FIRE_MODE_TOGGLE, initial(choice.fire_mode), user.client)
+
+	base_gun_icon = initial(choice.icon_state)
+	update_icon()
+
+	to_chat(user, initial(choice.message_to_user))
+
+	var/obj/screen/ammo/A = user.hud_used.ammo //The ammo HUD
+	A.update_hud(user)
+
+	return TRUE
+
+
+
+//Ammo/Charge functions
+/obj/item/weapon/gun/energy/lasgun/standard_marine_sniper/update_icon(mob/user)
+	var/cell_charge = (!cell || cell.charge <= 0) ? 0 : CEILING((cell.charge / max(cell.maxcharge, 1)) * 100, 25)
+	icon_state = "[base_gun_icon]_[cell_charge]"
+	update_mag_overlay(user)
+
+
+/obj/item/weapon/gun/energy/lasgun/standard_marine_sniper/update_item_state(mob/user)
+    . = item_state
+    item_state = "[initial(icon_state)][flags_item & WIELDED ? "_w" : ""]"
+    if(. != item_state && ishuman(user))
+        var/mob/living/carbon/human/human_user = user
+        if(src == human_user.l_hand)
+            human_user.update_inv_l_hand()
+        else if (src == human_user.r_hand)
+            human_user.update_inv_r_hand()
+
+
+//TE Standard ML //Yes Icon names are called TE-M/Machine gun, but I'm too lazy and i'm too sleepy to fix them
+
+obj/item/weapon/gun/energy/lasgun/standard_marine_mlaser
+	name = "\improper TE-ML Machine laser rifle"
```

Comment:
```suggestion
	name = "\improper TE-ML autolaser"
```
again, MJP said no, but do you think it applies here?

Reply:
autolaser makes no sense and would imply it has auto fire to be the thing that sets it apart, the guns should have their surnames be simple and understandable rather than babble like autolaser as they're gong to most likely be the name most people use to refer to them

Reply:
```suggestion
	name = "\improper TE-ML machine laser rifle"
```
"Machine" does not make any sense to me in a rapid-firing, surpressing role though.

Reply:
machinegun is probably a better term than machine laser rifle to be fair, as it is basically a machinegun in role, and is easy to absorb and understand how it works.